test/token/ERC20WithPermit.test.js

```diff
@@ -0,0 +1,1102 @@
+const {
+  increaseTime,
+  lastBlockTime,
+  to1e18,
+  MAX_UINT256,
+  ZERO_ADDRESS,
+} = require("../helpers/contract-test-helpers")
+
+describe("ERC20WithPermit", () => {
+  // default Hardhat's networks blockchain, see https://hardhat.org/config/
+  const hardhatNetworkId = 31337
+
+  const initialSupply = to1e18(100)
+
+  let owner
+  let initialHolder
+  let recipient
+  let anotherAccount
+
+  let token
+
+  beforeEach(async () => {
+    ;[owner, initialHolder, recipient, anotherAccount] =
+      await ethers.getSigners()
+
+    const ERC20WithPermit = await ethers.getContractFactory("ERC20WithPermit")
+    token = await ERC20WithPermit.deploy("My Token", "MT")
+    await token.deployed()
+
+    await token.mint(initialHolder.address, initialSupply)
+  })
+
+  it("should have a name", async () => {
+    expect(await token.name()).to.equal("My Token")
+  })
+
+  it("should have a symbol", async () => {
+    expect(await token.symbol()).to.equal("MT")
+  })
+
+  it("should have 18 decimals", async () => {
+    expect(await token.decimals()).to.equal(18)
+  })
+
+  describe("totalSupply", () => {
+    it("should return the total amount of tokens", async () => {
+      expect(await token.totalSupply()).to.equal(initialSupply)
+    })
+  })
+
+  describe("permit_typehash", () => {
+    it("should be keccak256 of EIP2612 Permit message", async () => {
+      const expected = ethers.utils.keccak256(
+        ethers.utils.toUtf8Bytes(
+          "Permit(address owner,address spender,uint256 value,uint256 nonce,uint256 deadline)"
+        )
+      )
+      expect(await token.PERMIT_TYPEHASH()).to.equal(expected)
+    })
+  })
+
+  describe("domain_separator", () => {
+    it("should be keccak256 of EIP712 domain struct", async () => {
+      const keccak256 = ethers.utils.keccak256
+      const defaultAbiCoder = ethers.utils.defaultAbiCoder
+      const toUtf8Bytes = ethers.utils.toUtf8Bytes
+
+      const expected = keccak256(
+        defaultAbiCoder.encode(
+          ["bytes32", "bytes32", "bytes32", "uint256", "address"],
+          [
+            keccak256(
+              toUtf8Bytes(
+                "EIP712Domain(string name,string version,uint256 chainId,address verifyingContract)"
+              )
+            ),
+            keccak256(toUtf8Bytes("My Token")),
+            keccak256(toUtf8Bytes("1")),
+            hardhatNetworkId,
+            token.address,
+          ]
+        )
+      )
+      expect(await token.DOMAIN_SEPARATOR()).to.equal(expected)
+    })
+  })
+
+  describe("balanceOf", () => {
+    context("when the requested account has no tokens", () => {
+      it("should return zero", async () => {
+        expect(await token.balanceOf(anotherAccount.address)).to.equal(0)
+      })
+    })
+
+    context("when the requested account has some tokens", () => {
+      it("should return the total amount of tokens", async () => {
+        expect(await token.balanceOf(initialHolder.address)).to.equal(
```

Comment:
Can we use another account that holds just a specific number of tokens and not the whole token supply?